lib/hanami/entity.rb

```diff
@@ -61,238 +51,83 @@ module Hanami
+      attributes.fetch(:id, nil)
     end
 
-    # Return the hash of attributes
-    #
-    # @since 0.2.0
+    # Handle dynamic accessors
     #
-    # @example
-    #   require 'hanami/model'
-    #   class User
-    #     include Hanami::Entity
-    #     attributes :name
-    #   end
+    # If internal attributes set has the requested key, it returns the linked
+    # value, otherwise it raises a <tt>NoMethodError</tt>
     #
-    #   user = User.new(id: 23, name: 'Luca')
-    #   user.to_h # => { :id => 23, :name => "Luca" }
-    def to_h
-      Hash[attribute_names.map { |a| [a, read_attribute(a)] }]
+    # @since x.x.x
+    def method_missing(m)
+      attributes.fetch(m) { super }
     end
 
-    # Return the set of attribute names
+    # Implement generic equality for entities
+    #
+    # Two entities are equal if they are instances of the same class and they
+    # have the same id.
     #
-    # @since 0.5.1
+    # @param other [Object] the object of comparison
     #
-    # @example
-    #   require 'hanami/model'
-    #   class User
-    #     include Hanami::Entity
-    #     attributes :name
-    #   end
+    # @return [FalseClass,TrueClass] the result of the check
     #
-    #   user = User.new(id: 23, name: 'Luca')
-    #   user.attribute_names # #<Set: {:id, :name}>
-    def attribute_names
-      self.class.attributes
+    # @since 0.1.0
+    def ==(other)
+      self.class == other.class &&
+        id == other.id
     end
 
-    # Return the contents of the entity as a nicely formatted string.
-    #
-    # Display all attributes of the entity for inspection (even if they are nil)
-    #
-    # @since 0.5.1
+    # Implement predictable hashing for hash equality
     #
-    # @example
-    #   require 'hanami/model'
-    #   class User
-    #     include Hanami::Entity
-    #     attributes :name, :email
-    #   end
+    # @return [Integer] the object hash
     #
-    #   user = User.new(id: 23, name: 'Luca')
-    #   user.inspect # #<User:0x007fa7eefe0b58 @id=nil @name="Luca" @email=nil>
-    def inspect
-      attr_list = attribute_names.inject([]) do |res, name|
-        res << "@#{name}=#{read_attribute(name).inspect}"
-      end.join(' ')
-
-      "#<#{self.class.name}:0x00#{(__id__ << 1).to_s(16)} #{attr_list}>"
+    # @since x.x.x
+    def hash
+      [self.class, id].hash
     end
 
-    alias_method :to_s, :inspect
-
-    # Set attributes for entity
-    #
-    # @since 0.2.0
+    # Serialize entity to a Hash
     #
-    # @example
-    #   require 'hanami/model'
-    #   class User
-    #     include Hanami::Entity
-    #     attributes :name
-    #   end
+    # @return [Hash] the result of serialization
     #
-    #   user = User.new(name: 'Lucca')
-    #   user.update(name: 'Luca')
-    #   user.name # => 'Luca'
-    def update(attributes={})
-      attributes.each do |attribute, value|
-        public_send("#{attribute}=", value)
-      end
+    # @since 0.1.0
+    def to_h
+      attributes.deep_dup.to_h
```

Comment:
@beauby there are no circular associations in rom (thus hanami-model). we're dealing with aggregates, so only root objects know about their children, and never the other way around.